Change scheduled reports to use time-based checkpointing

dannyroberts opened the pull request:
This isn't quite ready but I'm opening it for comments. @czue this is I think basically what you suggested this morning.

(I still have do some testing, the last commit hasn't been run by a machine yet.)

dannyroberts committed:
Change scheduled reports to use time-based checkpointing

This should fix the most glaring problem with scheduled reports where
they whole batches of reports just get dropped if there is a 5 minute
delay in the periodic queue.

dannyroberts committed:
Tighten up interface for _make_all_notification_view_keys

force caller to pass in a 15-minute aligned point in time

dannyroberts committed:
Add _iter_15_minute_marks_in_range

for future use in firing _all_ due scheduled reports

dannyroberts committed:
Strip pointless try except around generator function call

Tests assert that get_scheduled_report_ids raises a ValueError,
so not only does this try/except do nothing, but our code depends on it
not working (or at least our tests do). I'm opting to just remove it.

dannyroberts committed:
Pull _keys function out of get_scheduled_report_ids

dannyroberts committed:
Rename get_scheduled_report_ids keys helper and pull out guess_reporting_minute

dannyroberts committed:
Run makemigrations saved_reports

dannyroberts committed:
Remove unused import

dannyroberts committed:
Remove extra recursive call to daily_reports task

dannyroberts committed:
Remove calls to removed functions

czue commented on corehq/apps/saved_reports/scheduled.py:
```diff
@@ -49,11 +49,7 @@ def _keys(period, as_of):
-    except ValueError:
-        _soft_assert(False, "Celery was probably down for a while. Lots of reports are getting dropped!")
-        raise StopIteration
+    keys = _keys(period, as_of)
```
lol @ this commit message.

czue commented on corehq/apps/saved_reports/scheduled.py:
```diff
@@ -88,3 +88,22 @@ def guess_reporting_minute(now=None):
+        return rounded_down + timedelta(minutes=1)
+
+
+def _iter_15_minute_marks_in_range(start_datetime, end_datetime):
```
these two functions are screaming for some unit tests

dannyroberts replied:
Totally. I'll add some

czue reviewed (approved):
Yes! This is great - exactly what i was thinking.

dannyroberts commented on corehq/apps/saved_reports/tasks.py:
```diff
@@ -62,26 +63,13 @@ def send_report_throttled(notification_id):
-def weekly_reports():
-    for report_id in get_scheduled_report_ids('weekly'):
-        send_delayed_report(report_id)
+def initiate_queue_scheduled_reports():
```
This pattern seems a little silly/wasteful to me, but I saw it used elsewhere so I thought I'd start with it.

Essentially the lifecycle for firing scheduled reports would now be

```
beat => celery_periodic => initiate_queue_scheduled_reports() => celery_periodic => queue_scheduled_reports() (serial task)
```
followed by the rest of the existing process
```
... => background_queue (one per scheduled report) => send_report(notification_id)
```
I think the ideal thing were if it were possible to call `queue_scheduled_reports()` directly instead of `queue_scheduled_reports.delay()`, but I'm not sure whether that works. Anyone know?

My thinking about why this might matter is that if we have a backlog in the `celery_periodic` queue and the time-to-start for `celery_periodic` is 10 minutes, then with the double round trip it'll take 20 minutes before the actual send_report tasks are even queued. If no one has suggestions, I'm fine not optimizing around this in the first pass.

snopoke replied:
I don't see any reason why that wouldn't work. Agree that the way it is now seems odd.

dannyroberts replied:
OK I made a note to myself to revisit this after the feature's rolled out.

dannyroberts replied:
@czue are you suggesting that

```
@serial_task(...)
@periodic_task(...)
def queue_scheduled_reports():
    ...
```
would work? That doesn't seem right to me because both of those are a kind of `@task` decorator, so it'd be declared as a task in two different ways, and I don't have a mental model for how that'd work; I'm not even sure the behavior there is well defined.

What I was suggesting (and maybe what you were saying too) is to replace

```
@periodic_task(...)
def initiate_queue_scheduled_reports():
    queue_scheduled_reports.delay()
```
with
```
@periodic_task(...)
def initiate_queue_scheduled_reports():
    queue_scheduled_reports()  # no .delay here
```

What I would _want_ that to do is run `queue_scheduled_reports` directly without putting it back in the queue, but if serial_task finds another process is running `queue_scheduled_reports` at the same time, that it _do_ get put on the queue to be retried at a future date. This may be the actual behavior, but I thin my uncertainty comes from not knowing what `self.retry` does in the situation when a task function was called directly instead of through celery. It seems reasonable that what I want would happen there. It seems like a simple-enough thing to try, so maybe I'll just do a local test and see what happens.

dannyroberts replied:
Actually just to amend what I said, I'll test that the main case works, as a sanity check. There's no reason the locking should work any different, so two running at a time isn't a concern, and if retry doesn't work the worst that'll happen is that we get a celery error and we wait until the next 15-minute mark to fire again. That's a remarkably inconsequential downside, so I'll keep an eye out for such a message in sentry once we deploy, but I won't try to trigger that locally to see what happens.

czue replied:
Right, I see what you're saying in the first comment. And in the second you're just saying that making it `serial` isn't actually necessary because of locking, yeah? That also makes sense to me.

I think if for whatever reason we wanted to bring back the serial-ness granted by the `@serial_task` decorator, we could split it up into a `@serial` (that deals with locking and such) and `@task` (that actually makes it a task)? Basically extracting the contents of `[_inner](https://github.com/dimagi/commcare-hq/blob/master/corehq/util/decorators.py#L121-L135)` and making them a separate decorator and then having `@serial_task` and this code use that. 

But the above doesn't seem necessary in this case since as you pointed out we're already managing locking via the DB.

dannyroberts replied:
Oh I wasn't even factoring in db level locking, which I'm not super confident about/haven't thought through. I was referring to what I eventually did in this commit: https://github.com/dimagi/commcare-hq/pull/23737/commits/36b9e3d8a4398df8245ed91f86e59c7f70f90889, i.e.

```
@periodic_task(...)
def initiate_queue_scheduled_reports():
    queue_scheduled_reports()  # no .delay here
```

`queue_scheduled_reports` is still a `@serial_task`. My [second comment](https://github.com/dimagi/commcare-hq/pull/23737#discussion_r270539567) was really just saying that I realized the details of how it works in the retry case (when it can't acquire the redis lock) weren't actually that important, since we don't really depend on the retry anyway.

What you said about splitting `@serial_task` into `@serial` and the existing `@task` makes sense, though I think you'd have to be tricky to get the same retry behavior for a serial task in that case.

Anyways, I think I've made the change I wanted to make that came out of this convo.

czue replied:
> Anyways, I think I've made the change I wanted to make that came out of this convo.

 Sounds good. Sorry if I confused things...

dannyroberts committed:
Add tests for _iter_15_minute_marks_in_range and _round_datetime_up_to_the_nearest_minute

dannyroberts committed:
Fix scheduled reports tests for get_scheduled_report_ids's new boundary behavior

It used to be that get_scheduled_report_ids would accept an exact minute
mark for its "as_of" argument, but this has been transformed into an
end_datetime arugment which is exclusive. It is now required to be
stricly greater than the 15-minute mark (or 15-minute marks) that are
being targetted. I fixed this by changing a test helper method to add a
tiny fudge factor to the as_of datetime it receives.

dannyroberts committed:
Fix stickler error

dannyroberts committed:
Merge branch 'dmr/pull-saved-reports-into-separate-app' into dmr/add-scheduled-reports-checkpoint

dannyroberts committed:
Fix new scheduled report checkpoint logic after local testing

dannyroberts committed:
Merge branch 'dmr/pull-saved-reports-into-separate-app' into dmr/add-scheduled-reports-checkpoint

dannyroberts committed:
Remove nondeterministic scheduled reports test testDefaultValue

dannyroberts committed:
Move cleanup in scheduled reports tests to tearDown

dannyroberts committed:
Add test for create_records_for_scheduled_reports

dannyroberts committed:
Fix stickler errors for saved_exports checkpointing tests

dannyroberts committed:
Merge branch 'master' into dmr/add-scheduled-reports-checkpoint

snopoke commented on corehq/apps/saved_reports/models.py:
```diff
@@ -702,3 +704,15 @@ def update_attributes(self, items):
             raise ValidationError("You can not specify a start date in the past.")
+
+
+class ScheduledReportsCheckpoint(models.Model):
```
could do with a docstring to outline it's purpose

dannyroberts replied:
Thanks, will add

dannyroberts committed:
Describe ScheduledReportsCheckpoint's role in a docstring

dannyroberts commented:
Just tested this on staging, and it behaved as expected. I think I'm ready for a final approve and then to roll this out

czue reviewed (commented):
just realized i commented several days ago but never completed the review. but agree with simon

dannyroberts committed:
Have initiate_queue_scheduled_reports call queue_scheduled_reports directly

Currently initiate_queue_scheduled_reports calls queue_scheduled_reports.delay,
which puts it back on the periodic_queue. This means that we need to wait
two round trips on the periodic_queue before we even start running any code
related to scheduled reports. This cuts it down to a single round trip on the periodic_queue.

dannyroberts committed:
Merge branch 'master' into dmr/add-scheduled-reports-checkpoint

dannyroberts commented:
Actually I see the most recent build before we started having travis issues passed, so I'm inclined to merge this in for today's deploy. Seeking a final approval

dannyroberts committed:
Merge branch 'master' into dmr/add-scheduled-reports-checkpoint
